pkg/reconciler/service/service.go

```diff
@@ -273,28 +273,38 @@ func (c *Reconciler) checkRoutesNotReady(config *v1alpha1.Configuration, logger
 	}
 }
 
-func (c *Reconciler) updateStatus(desired *v1alpha1.Service, logger *zap.SugaredLogger) (*v1alpha1.Service, error) {
-	service, err := c.serviceLister.Services(desired.Namespace).Get(desired.Name)
-	if err != nil {
-		return nil, err
-	}
-	// If there's nothing to update, just return.
-	if reflect.DeepEqual(service.Status, desired.Status) {
-		return service, nil
-	}
-	becomesReady := desired.Status.IsReady() && !service.Status.IsReady()
-	// Don't modify the informers copy.
-	existing := service.DeepCopy()
-	existing.Status = desired.Status
-
-	svc, err := c.ServingClientSet.ServingV1alpha1().Services(desired.Namespace).UpdateStatus(existing)
-	if err == nil && becomesReady {
-		duration := time.Since(svc.ObjectMeta.CreationTimestamp.Time)
-		logger.Infof("Service became ready after %v", duration)
-		c.StatsReporter.ReportServiceReady(service.Namespace, service.Name, duration)
-	}
+func (c *Reconciler) updateStatus(desired *v1alpha1.Service, logger *zap.SugaredLogger) error {
+	return reconciler.RetryUpdateConflicts(func(i int) error {
+		var (
+			existing *v1alpha1.Service
+			err      error
+		)
+		// The first iteration tries to use the informer's state.
+		if i == 0 {
+			existing, err = c.serviceLister.Services(desired.Namespace).Get(desired.Name)
+			existing = existing.DeepCopy()
+		} else {
+			existing, err = c.ServingClientSet.ServingV1alpha1().Services(desired.Namespace).Get(desired.Name, metav1.GetOptions{})
+		}
+		if err != nil {
+			return err
+		}
 
-	return svc, err
+		// If there's nothing to update, just return.
+		if reflect.DeepEqual(existing.Status, desired.Status) {
+			return nil
+		}
+
+		becomesReady := desired.Status.IsReady() && !existing.Status.IsReady()
```

Comment:
This makes me wonder - what's causing the conflicts? (ie. is it the labeller or gc'er)